controllers/cluster_controller_phases.go

```diff
@@ -138,27 +138,32 @@ func (r *ClusterReconciler) reconcileInfrastructure(ctx context.Context, cluster
 		return err
 	}
 
-	if cluster.Status.InfrastructureReady || !infraConfig.GetDeletionTimestamp().IsZero() {
+	// There's no need to go any further if the Cluster is marked for deletion.
+	if !infraConfig.GetDeletionTimestamp().IsZero() {
 		return nil
 	}
 
 	// Determine if the infrastructure provider is ready.
-	ready, err := external.IsReady(infraConfig)
-	if err != nil {
-		return err
-	} else if !ready {
-		klog.V(3).Infof("Infrastructure provider for Cluster %q in namespace %q is not ready yet", cluster.Name, cluster.Namespace)
-		return nil
+	if !cluster.Status.InfrastructureReady {
+		ready, err := external.IsReady(infraConfig)
+		if err != nil {
+			return err
+		} else if !ready {
+			klog.V(3).Infof("Infrastructure provider for Cluster %q in namespace %q is not ready yet", cluster.Name, cluster.Namespace)
+			return nil
+		}
+		cluster.Status.InfrastructureReady = true
 	}
-	cluster.Status.InfrastructureReady = true
 
 	// Get and parse Status.APIEndpoint field from the infrastructure provider.
-	if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Status.APIEndpoints, "status", "apiEndpoints"); err != nil {
-		return errors.Wrapf(err, "failed to retrieve Status.APIEndpoints from infrastructure provider for Cluster %q in namespace %q",
-			cluster.Name, cluster.Namespace)
-	} else if len(cluster.Status.APIEndpoints) == 0 {
-		return errors.Wrapf(err, "retrieved empty Status.APIEndpoints from infrastructure provider for Cluster %q in namespace %q",
-			cluster.Name, cluster.Namespace)
+	if len(cluster.Status.APIEndpoints) == 0 {
```

Comment:
@akutz @vincepri Any particular reason here to only set these once? It seems like it might be better to attempt to update these every time, otherwise I'm not sure the upgrade workflow would work for a provider that does not use a load balancer.

Reply:
So that's what I would do, but it's not what was done before. I was attempting minimal impact. I've been writing code that reconciles, but I've noted a lot of CAPI code is more about initialization, so I tend to stick with the practice in the repo for which I'm writing the PR.

Reply:
A possible side-effect, though, is if the control-plane endpoint is taken from an API server endpoint and later the API server list changes and the certificates no longer match the kubeconfig generated originally.

Reply:
The contract around `Status.Ready` assumes immutability. Once a provider sets the value to `true` it should never go back to `false` from a Cluster API point of view. There is no such contract for the other values though, if we allow for APIEndpoints to be updated, that should be fine.

Regarding the change, I'm not sure what `a bug where machines cannot depend on a cluster's InfrastructureReady` actually means?

Reply:
I need to go remind myself. I also don’t think immutability should be assumed, but okay.

FWIW, this repo’s tests fail with the other order as well. It wasn’t just CAPV.

Reply:
Oh yeah, I meant that if the API endpoint check is first, and it returns an error until there are API endpoints set, then this causes an ordering bug. Because InfraReady is never set on the CAPI Cluster, and if a machine reconciler is waiting for InfraReady, then it will never reconcile, potentially never supplying the needed API endpoints.

Reply:
For the kubeconfig, we likely will need to:
- Determine if we created it vs created externally (ownerref maybe)?
- If we created it and if the defined endpoint is not == apiendpoints[0], recreate it

Reply:
Infrastructure provider machines shouldn't rely on anything other than Bootstrap.Data to be available to start spinning up instances. 

Kubeadm provider does all the heavy lifting here waiting for the `InfrastructureReady` and for 
 `ControlPlaneInitialized`

Reply:
@vincepri It gets tricky without a stable control plane endpoint, which is the case with CAPV today. We'll need to ensure that the API endpoints get updated and that the generated kubeconfig is updated as control plane machines are replaced, otherwise the result will be breakage.

Reply:
@detiber That sounds good to me, we can do 1, but I think we should hold on for 2 until after release, wdyt?

Reply:
I think this PR fixes CAPV today, and you don’t want the API endpoint error to prevent InfraReady from being set, which was the original intent of your first change anyway.

Reply:
And it's also the case that the kubeadm bootstrap provider checks for InfraReady before creating bootstrap data. If InfraReady is never set because reading API endpoints causes an error, then the bootstrap data is never generated.